operators/constellation-node-operator/internal/executor/executor.go

```diff
@@ -0,0 +1,164 @@
+
+// Controller is a type with a reconcile method.
+// It is modeled after the controller-runtime reconcile method,
+// but reconciles on external resources instead of k8s resources.
+type Controller interface {
+	Reconcile(ctx context.Context) (Result, error)
+}
+
+// Executor is a task executor / scheduler.
+// It will call the reconcile method of the given controller with a regular interval
+// or when triggered externally.
+type Executor struct {
+	mux     sync.Mutex
+	started bool
+
+	// controller is the controller to be reconciled.
+	controller Controller
+
+	// pollingFrequency is the default frequency with which the controller is reconciled
+	// if no external trigger is received and no requeue is requested by the controller.
+	pollingFrequency time.Duration
+	// externalTrigger is used to trigger a reconciliation immediately from the outside.
+	// multiple triggers in a short time will be coalesced into one externalTrigger.
+	externalTrigger chan struct{}
+	// stop is used to stop the reconciliation loop.
+	stop chan struct{}
+}
+
+// New creates a new Executor.
+func New(controller Controller, cfg Config) *Executor {
+	cfg.applyDefaults()
+	return &Executor{
+		controller:       controller,
+		pollingFrequency: cfg.PollingFrequency,
+	}
+}
+
+// Start starts the executor in a separate go routine.
+// Call Stop to stop the executor.
+func (e *Executor) Start(ctx context.Context, wg *sync.WaitGroup) {
+	e.mux.Lock()
+	defer e.mux.Unlock()
+	if e.started {
+		return
+	}
+	e.started = true
+
+	logr := log.FromContext(ctx)
+	e.externalTrigger = make(chan struct{}, 1)
+	e.stop = make(chan struct{}, 1)
+	// execute is used by the go routines below to communicate
+	// that a reconciliation should happen
+	execute := make(chan struct{}, 1)
+	// nextScheduledReconcile is used to communicate the next scheduled reconciliation time
+	nextScheduledReconcile := make(chan time.Duration, 1)
+	// trigger a reconciliation on startup
+	nextScheduledReconcile <- 0
+
+	wg.Add(2)
+
+	// timer routine is responsible for triggering the reconciliation after the timer expires
+	// or when triggered externally
+	go func() {
+		defer wg.Done()
+		defer close(execute)
+		defer logr.Info("Timer stopped")
+		for {
+			nextScheduledReconcileAfter := <-nextScheduledReconcile
+			timer := *time.NewTimer(nextScheduledReconcileAfter)
+			select {
+			case <-e.stop:
+				return
+			case <-ctx.Done():
+				return
+			case <-e.externalTrigger:
+			case <-timer.C:
+			}
+			execute <- struct{}{}
+		}
+	}()
+
+	// executor routine is responsible for executing the reconciliation
+	go func() {
+		defer wg.Done()
+		defer close(nextScheduledReconcile)
+		defer logr.Info("Executor stopped")
+
+		for {
+			_, ok := <-execute
+			// execute channel closed. executor should stop
+			if !ok {
+				return
+			}
+			res, err := e.controller.Reconcile(ctx)
+			if err != nil {
+				logr.Error(err, "reconciliation failed")
+				nextScheduledReconcile <- 0 // requeue immediately
```

Comment:
should we limit the number of failure retries to avoid an infinite retry loop?

Reply:
This uses the same retry logic as kubernetes operators: https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/reconcile#Reconciler.
While looking into this a bit more, I learned that the controller-runtime offers configurable rate limiting strategies:

- https://github.com/kubernetes-sigs/controller-runtime/blob/a784ee78d04bdf216006adfbc313328a0aded88a/pkg/controller/controller.go#L57-L60
- https://github.com/kubernetes-sigs/controller-runtime/blob/a784ee78d04bdf216006adfbc313328a0aded88a/pkg/ratelimiter/ratelimiter.go#L21-L30
- https://pkg.go.dev/k8s.io/client-go/util/workqueue?utm_source=godoc#RateLimiter
- https://github.com/kubernetes/client-go/blob/master/util/workqueue/default_rate_limiters.go#L77-L83

The default seems to be exponential backoff with a cap of retrying once every 1000 seconds (~16 minutes).

So while I wouldn't want to set a limit, using a rate limiting strategy (preferably exponential backoff) sounds very reasonable. Wdyt?

Reply:
i thought of exponential backoff as first solution too. If not too hard to implement using the default k8s limiter and adding a check rateLimiter.When(struct{}) before requeueing seems enough right?